File: python-examples/bs4-example/api/list.py

from bs4 import BeautifulSoup
from intuned_browser import go_to_url
from intuned_runtime import extend_payload
from playwright.async_api import BrowserContext, Page
from utils.types_and_schemas import ListParams, Product


def extract_products(html: str) -> list[Product]:
    """
    Extracts product info from the page HTML using BeautifulSoup.
    Replace selectors with appropriate ones for your target site.
    """
    soup = BeautifulSoup(html, "html.parser")
    products: list[Product] = []

    # Find all product items - replace selector as needed
    product_elements = soup.select("li.product")

    for product in product_elements:
        # Extract title - replace selector as needed
        title_element = product.select_one("h2.woocommerce-loop-product__title")
        title = title_element.get_text(strip=True) if title_element else ""

        # Extract price - replace selector as needed
        price_element = product.select_one("span.woocommerce-Price-amount")
        price = price_element.get_text(strip=True) if price_element else ""

        # Extract details URL - replace selector as needed
        link_element = product.select_one("a.woocommerce-LoopProduct-link")
        details_url = link_element.get("href", "") if link_element else ""

        if title and details_url:
            products.append(
                Product(
                    title=title,
                    price=price,
                    details_url=details_url,
                )
            )

    return products


def get_next_page_url(html: str) -> str | None:
    """
    Extracts the next page URL from pagination.
    Returns None if there is no next page.
    """
    soup = BeautifulSoup(html, "html.parser")

    # Find the next page link - replace selector as needed
    next_link = soup.select_one("a.next.page-numbers")

    if next_link:
        return next_link.get("href")

    return None


async def automation(
    page: Page,
    params: ListParams,
    context: BrowserContext | None = None,
    **_kwargs,
):
    """
    Scrapes product listings from an e-commerce site using BeautifulSoup.
    Handles pagination by following "Next" page links.

    Example params:
    {
        "url": "https://www.scrapingcourse.com/ecommerce/",
        "max_pages": 5
    }

    This function:
    1. Navigates to the listing page
    2. Extracts product data using BeautifulSoup
    3. Follows pagination links to scrape multiple pages
    4. Returns all products found
    """
    params = ListParams(**params)
    # Set default values
    url = params.url
    if not url:
        raise ValueError("url is required in params")
    max_pages = params.max_pages

    print(f"Starting scrape from: {url}")
    print(f"Max pages: {max_pages}")

    all_products: list[Product] = []
    current_url: str | None = url
    current_page = 1

    while current_url and current_page <= max_pages:
        print(f"Scraping page {current_page}: {current_url}")

        # Navigate to the page
        await go_to_url(page, current_url)

        # Wait for products to load - replace selector as needed
        await page.wait_for_selector("li.product")

        # Get the page HTML
        html = await page.content()

        # Extract products using BeautifulSoup
        products = extract_products(html)
        all_products.extend(products)

        print(f"Found {len(products)} products on page {current_page}")

        # Get next page URL
        current_url = get_next_page_url(html)
        current_page += 1

    print(f"Total products scraped: {len(all_products)}")
    for product in all_products:
        extend_payload(
            {
                "api": "details",
                "parameters": product,
            }
        )
    return {
        "products": all_products,
        "count": len(all_products),
        "pages_scraped": current_page - 1,
    }

File: typescript-examples/jsdom/api/list.ts
import { BrowserContext, Page } from "playwright";
import { goToUrl } from "@intuned/browser";
import { extendPayload } from "@intuned/runtime";
import { JSDOM } from "jsdom";
import { z } from "zod";
import { listSchema, Product } from "../utils/typesAndSchemas.js";

type Params = z.infer<typeof listSchema>;

function extractProducts(html: string): Product[] {
  /**
   * Extracts product info from the page HTML using JSDOM.
   * Replace selectors with appropriate ones for your target site.
   */
  const dom = new JSDOM(html);
  const document = dom.window.document;
  const products: Product[] = [];

  // Find all product items - replace selector as needed
  const productElements = document.querySelectorAll("li.product");

  productElements.forEach((product) => {
    // Extract title - replace selector as needed
    const titleElement = product.querySelector(
      "h2.woocommerce-loop-product__title"
    );
    const title = titleElement?.textContent?.trim() || "";

    // Extract price - replace selector as needed
    const priceElement = product.querySelector("span.woocommerce-Price-amount");
    const price = priceElement?.textContent?.trim() || "";

    // Extract details URL - replace selector as needed
    const linkElement = product.querySelector("a.woocommerce-LoopProduct-link");
    const detailsUrl = linkElement?.getAttribute("href") || "";

    if (title && detailsUrl) {
      products.push({
        title,
        price,
        detailsUrl,
      });
    }
  });

  return products;
}

function getNextPageUrl(html: string): string | null {
  /**
   * Extracts the next page URL from pagination.
   * Returns null if there is no next page.
   */
  const dom = new JSDOM(html);
  const document = dom.window.document;

  // Find the next page link - replace selector as needed
  const nextLink = document.querySelector("a.next.page-numbers");

  if (nextLink) {
    return nextLink.getAttribute("href");
  }

  return null;
}

export default async function handler(
  params: Params,
  page: Page,
  context: BrowserContext
) {
  /**
   * Scrapes product listings from an e-commerce site using JSDOM.
   * Handles pagination by following "Next" page links.
   *
   * Example params:
   * {
   *   "url": "https://www.scrapingcourse.com/ecommerce/",
   *   "maxPages": 5
   * }
   *
   * This function:
   * 1. Navigates to the listing page
   * 2. Extracts product data using JSDOM
   * 3. Follows pagination links to scrape multiple pages
   * 4. Returns all products found
   */

  // Set default values
  const url = params.url;
  if (!url) {
    throw new Error("url is required in params");
  }
  const maxPages = params.maxPages ?? 10;

  console.log(`Starting scrape from: ${url}`);
  console.log(`Max pages: ${maxPages}`);

  const allProducts: Product[] = [];
  let currentUrl: string | null = url;
  let currentPage = 1;

  while (currentUrl && currentPage <= maxPages) {
    console.log(`Scraping page ${currentPage}: ${currentUrl}`);

    // Navigate to the page
    await goToUrl({ page, url: currentUrl });

    // Wait for products to load - replace selector as needed
    await page.waitForSelector("li.product");

    // Get the page HTML
    const html = await page.content();

    // Extract products using JSDOM
    const products = extractProducts(html);
    allProducts.push(...products);

    console.log(`Found ${products.length} products on page ${currentPage}`);

    // Get next page URL
    currentUrl = getNextPageUrl(html);
    currentPage += 1;
  }

  console.log(`Total products scraped: ${allProducts.length}`);

  for (const product of allProducts) {
    extendPayload({
      api: "details",
      parameters: product,
    });
  }

  return {
    products: allProducts,
    count: allProducts.length,
    pagesScraped: currentPage - 1,
  };
}
